browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,120 @@
+#include "chrome/browser/ui/simple_message_box.h"
+#include "chrome/common/pref_names.h"
+#include "chrome/grit/chromium_strings.h"
+#include "components/country_codes/country_codes.h"
+#include "components/grit/brave_components_strings.h"
+#include "components/prefs/pref_service.h"
+#include "net/dns/public/dns_over_https_config.h"
+#include "net/dns/public/doh_provider_entry.h"
+#include "net/dns/public/secure_dns_mode.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace secure_dns = chrome_browser_net::secure_dns;
+
+namespace brave_vpn {
+
+namespace {
+std::string GetFilteredProvidersForCountry() {
+  // Use default servers for current country.
+  auto providers = secure_dns::ProvidersForCountry(
+      secure_dns::SelectEnabledProviders(net::DohProviderEntry::GetList()),
+      country_codes::GetCurrentCountryID());
+  std::string doh_https_templates;
+  for (const auto* entry : providers) {
+    if (!doh_https_templates.empty()) {
+      doh_https_templates += " ";
+    }
+    net::DnsOverHttpsConfig doh_config({entry->doh_server_config});
+    doh_https_templates += doh_config.ToString();
+  }
+  return doh_https_templates;
+}
+std::string GetDoHServers(SecureDnsConfig* dns_config) {
+  return dns_config && !dns_config->doh_servers().servers().empty()
+             ? dns_config->doh_servers().ToString()
+             : GetFilteredProvidersForCountry();
+}
+}  // namespace
+
+BraveVpnDnsObserverService::BraveVpnDnsObserverService(PrefService* local_state)
+    : local_state_(local_state) {
+  pref_change_registrar_.Init(local_state);
+  pref_change_registrar_.Add(
+      prefs::kDnsOverHttpsMode,
+      base::BindRepeating(&BraveVpnDnsObserverService::OnDNSPrefChanged,
+                          weak_ptr_factory_.GetWeakPtr()));
+}
+
+BraveVpnDnsObserverService::~BraveVpnDnsObserverService() = default;
+
+bool BraveVpnDnsObserverService::ShouldAllowExternalChanges() const {
+  if (allow_changes_for_testing_.has_value())
+    return allow_changes_for_testing_.value();
+  auto* browser = chrome::FindLastActive();
+  return (chrome::ShowQuestionMessageBoxSync(
+              browser ? browser->window()->GetNativeWindow()
+                      : gfx::kNullNativeWindow,
+              l10n_util::GetStringUTF16(IDS_PRODUCT_NAME),
+              l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_CHANGE_ALERT)) ==
+          chrome::MESSAGE_BOX_RESULT_YES);
+}
+
+void BraveVpnDnsObserverService::OnDNSPrefChanged() {
+  if (!active_)
+    return;
+  // Reset saved config and keep user's choice.
+  if (ShouldAllowExternalChanges()) {
+    user_dns_config_.reset();
+  } else {
+    active_ = false;
+    SetDNSOverHTTPSMode(SecureDnsConfig::kModeSecure,
+                        GetDoHServers(user_dns_config_.get()));
+    active_ = true;
+  }
+}
+
+void BraveVpnDnsObserverService::SetDNSOverHTTPSMode(
+    const std::string& mode,
+    const std::string& doh_providers) {
+  local_state_->SetString(prefs::kDnsOverHttpsTemplates, doh_providers);
```

Comment:
Something we might want to try (I can help test this tomorrow). Checking behavior when group policy is overriding these values. For example, see:
- https://chromeenterprise.google/policies/#DnsOverHttpsTemplates
- https://chromeenterprise.google/policies/#DnsOverHttpsMode

Basically, if person has admin policy for `DnsOverHttpsMode` as `off`, we might not want to create the observer and instead we can show a warning (maybe security/privacy review can have guidance)

Reply:
with policy [DnsOverHttpsMode](https://chromeenterprise.google/policies/#DnsOverHttpsMode) off we have DNS leak. The settings page blocks access to the option in this case. The changes from observer do not have effect, policies are applied later on reading. Good idea to show another message about it to user.

Reply:
Moved warning message to a new issue https://github.com/brave/brave-browser/issues/23123

Reply:
if Brave crashes or vpn is left on when the browser shuts down we're going to lose the user's dns config

Reply:
also what happens if the user sets the dns config after vpn is enabled?

Reply:
As shared in DM, we should be good to go here 👍 I did testing tonite to verify the behavior
- VPN is system wide (using Windows or macOS); not owned by Brave process (ex: not socks server)
- VPN config stored with OS
- DNS over HTTP (with this PR) can only have two states
    1. **User did not set it yet**. If this is the case, turning on VPN sets Cloudflare and disabling turns off. If user picks another DoH provider, it keeps that new provider and turning VPN off will keep the DoH provider enabled.
    2. **User did set DoH**. If this is the case, the setting is not touched at all.